### ISSUE TITLE
What is the function of tensor "denom" ?

### ISSUE THREAD

**Issue by qwe1997** [Cartus/AGGCN_TACRED#38]
In the Class of  GraphConvLayer and MultiGraphConvLayer, there is a tensor called denom. I have tried to understand its meaning,however I failed.So please tell me this tensor's meaning.thanks!

` def forward(self, adj, gcn_inputs):
        # gcn layer
        denom = adj.sum(2).unsqueeze(2) + 1  
        outputs = gcn_inputs
        cache_list = [outputs]
        output_list = []

        for l in range(self.layers):  # 0, 1
            Ax = adj.bmm(outputs)
            AxW = self.weight_list[l](Ax)
            AxW = AxW + self.weight_list[l](outputs)  # self loop
            AxW = AxW / denom 
            gAxW = F.relu(AxW)
            cache_list.append(gAxW)
            outputs = torch.cat(cache_list, dim=2) 
            output_list.append(self.gcn_drop(gAxW))`

**Comment by Cartus** [comment]
As I remember, denom is the diagonal node degree matrix used in GCN, please refer to the original GCN paper:
https://arxiv.org/pdf/1609.02907.pdf

**Comment by qwe1997** [comment]
Thank you very much! "denom" is the diagonal node degree matrix, "adj.sum(2)" means getting graph nodes degrees and unsqueeze(2) means generating two dimensional matrix for graph convolution calculation and +1 means adding self-loop nodes degrees!

**Comment by Cartus** [comment]
Exactly! I am closing this issue, if you have further questions, feel free to re-open it.
